### PR TITLE
Bugfix #8116 - Unable to access store key in onPrepare and beforeSession hooks

### PR DESCRIPTION
**Pull request by Staicul**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)
Fix for bug https://github.com/webdriverio/webdriverio/issues/8116.
Sharing the server port with the service is now done via capabilities.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Review by christian-bromann — COMMENTED** [review]
Good looks good, thank you for the work 👍 

One comment: can we remove `packages/wdio-shared-store-service/src/utils.ts` as it seems we don't need it anymore

**Comment by Staicul** [comment]
Updated. Thanks for the support @christian-bromann !

File: packages/wdio-shared-store-service/tests/launcher.test.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { writeFile, deleteFile } from '../src/utils'
+import type { SharedStoreServiceCapabilities } from '../build/types'
```
Comment (christian-bromann) [inline packages/wdio-shared-store-service/tests/launcher.test.ts:1]: better to directly pull the type from the TypeScript file
```suggestion
import type { SharedStoreServiceCapabilities } from '../src/types'
```

**Comment by ruslan-qualtie** [comment]
@christian-bromann @Staicul 
Lines of code below have negative side effect.
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/wdio-shared-store-service/src/launcher.ts#L21-L23
If in `capabilities` used one of types (or similar)
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/wdio-types/src/Capabilities.ts#L478-L481
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/wdio-types/src/Capabilities.ts#L363-L366
Check below will not pass
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/webdriver/src/utils.ts#L35-L54
It will pass only If in `capabilities` used one of types (or similar)
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/wdio-types/src/Capabilities.ts#L590-L595
https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/wdio-types/src/Capabilities.ts#L393-L405
I have seen it in my work.
This limits the use of types in configuration, when `webdriverio` work with `shared-store-service` version 7.19.4 and above.
I don't know how to get rid of this side effect other than to freeze the `shared-store-service` version on 7.19.3

**Comment by christian-bromann** [comment]
@ruslan-qualtie can you please describe the side effect? The above doesn't make much sense to me. Thank you!

**Comment by ruslan-qualtie** [comment]
@christian-bromann @Staicul
> @ruslan-qualtie can you please describe the side effect?
Yes of course.

I have `wdioConfig.ts` created by `yarn create wdio`
I have two test app: `ios` and `android` .
For each app I construct `capabilities` in separate config file ( `iosConfig.ts` and  `androidConfig.ts` below).
I launch testss using `yarn wdio androidConfig.ts` and `yarn wdio iosConfig.ts`

`iosConfig.ts` :
```
import {
  DesiredCapabilities,
  AppiumW3CCapabilities,
  AppiumXCUITestCapabilities
} from '@wdio/types/build/Capabilities';
import path from 'path';
import { config } from './wdioConfig';

const appiumW3CCapabilities: AppiumW3CCapabilities = {
  'appium:automationName': 'XCUITest',
  'appium:platformName': 'iOS',
  'appium:platformVersion': '15.4',
  'appium:deviceName': 'iPhone 13 Pro',
  'appium:app': path.resolve('./apps/MyTest.app'),
  'appium:newCommandTimeout': 600,
  'appium:noReset': false
};

const appiumXCUITestCapabilities: AppiumXCUITestCapabilities = {
  'appium:bundleId': 'com.my.app.bundle.id',
  'appium:isHeadless': false,
  'appium:useXctestrunFile': true,
  'appium:bootstrapPath': path.resolve('./apps/WebDriverAgent/Build/Products')
};

const desiredCapabilities: DesiredCapabilities = {
  ...appiumW3CCapabilities,
  ...appiumXCUITestCapabilities
};

config.capabilities = [desiredCapabilities];

export { config };
```
`androidConfig.ts` :
```
import {
  DesiredCapabilities,
  Capabilities,
  AppiumCapabilities,
  AppiumAndroidCapabilities
} from '@wdio/types/build/Capabilities';
import path from 'path';
import { config } from './wdioConfig';

const capabilities: Capabilities = {
  platformName: 'Android'
};

const appiumCapabilities: AppiumCapabilities = {
  automationName: 'UiAutomator2',
  platformVersion: '12.0',
  deviceName: 'Pixel 3 API 31',
  app: path.resolve('./apps/MyTest.apk'),
  newCommandTimeout: 60,
  noReset: false
};

const appiumAndroidCapabilities: AppiumAndroidCapabilities = {
  androidInstallTimeout: 120000,
  isHeadless: false
};

const desiredCapabilities: DesiredCapabilities = {
  ...capabilities,
  ...appiumCapabilities,
  ...appiumAndroidCapabilities
};

config.capabilities = [desiredCapabilities];

export { config };
```

When I use  `shared-store-service` version 7.19.3 - both configs is working as expected.

When I use  `shared-store-service` version 7.19.4 or above  - only `ios` config work as expected.

Android tests fail launch with error:
```
[0-0] 2022-07-26T14:00:30.407Z INFO webdriver: Initiate new session using the WebDriver protocol
[0-0] 2022-07-26T14:00:30.409Z ERROR @wdio/runner: Error:
Invalid or unsupported WebDriver capabilities found ("automationName", "platformVersion", "deviceName", "app", "newCommandTimeout", "noReset", "androidInstallTimeout", "isHeadless"). 
Ensure to only use valid W3C WebDriver capabilities (see https://w3c.github.io/webdriver/#capabilities).
If you run your tests on a remote vendor, like Sauce Labs or BrowserStack, make sure that you put them 
into vendor specific capabilities, e.g. "sauce:options" or "bstack:options". Please reach out to 
to your vendor support team if you have further questions.

```

Error rise there:

https://github.com/webdriverio/webdriverio/blob/7167dcccfa22080acaec3fffcd94236110c23b5b/packages/webdriver/src/utils.ts#L45-L54

In order for the `androidСonfig.ts` to work with `shared-store-service` version 7.19.4 or above , I have to change it as shown below:

```
import {
  DesiredCapabilities,
  Capabilities,
  AppiumW3CCapabilities
} from '@wdio/types/build/Capabilities';
import path from 'path';
import { config } from './wdioConfig';

const capabilities: Capabilities = {
  platformName: 'Android'
};

const appiumW3CCapabilities: AppiumW3CCapabilities = {
  'appium:automationName': 'UiAutomator2',
  'appium:platformVersion': '12.0',
  'appium:deviceName': 'Pixel 3 API 31',
  'appium:app': path.resolve('./apps/MyTest.apk'),
  'appium:newCommandTimeout': 60,
  'appium:noReset': false
};

const desiredCapabilities: DesiredCapabilities = {
  ...capabilities,
  ...appiumW3CCapabilities
};

config.capabilities = [desiredCapabilities];

export { config };
```

But in this case I cannot use `AppiumAndroidCapabilities` type for construct `capabilities` in `androidConfig.ts`

**Comment by christian-bromann** [comment]
I believe all these caps can be defined to be valid w3c caps, e.g. `appium:androidInstallTimeout`. However we don't have typings for that. Mind raising an issue, and consider collaborating?

**Comment by ruslan-qualtie** [comment]
@christian-bromann 
Unfortunately I am now full time on the project.
I won't be able to find the time to collaborating.
Thanks for the tip to try use `appium:androidInstallTimeout`.
I'll try extension of type  or direct injection.

**Pull request merged**